From 04bb673333ae9bb7b741a2da42cdf65128452787 Mon Sep 17 00:00:00 2001
From: Zhihai Xu <zhihaixu2012@gmail.com>
Date: Tue, 7 Jun 2016 10:29:51 -0700
Subject: [PATCH 0770/1164] CDH-45293: HIVE-13760: Add a HIVE_QUERY_TIMEOUT
 configuration to kill a query if a query is
 running for more than the configured timeout
 value. (Zhihai Xu via Chao Sun)

Conflicts:
	common/src/java/org/apache/hadoop/hive/conf/HiveConf.java

Change-Id: Iae560d90f4b842bde78919bc0735a2f5b3bd19d5
---
 .../java/org/apache/hadoop/hive/conf/HiveConf.java |    6 +++++-
 .../hive/service/cli/operation/SQLOperation.java   |    6 ++++++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
index cd6fdc2..260db05 100644
--- a/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
+++ b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
@@ -2157,7 +2157,11 @@ public void setSparkConfigUpdated(boolean isSparkConfigUpdated) {
             "Comma-separated list of supported blobstore schemes."),
 
     HIVE_BLOBSTORE_USE_BLOBSTORE_AS_SCRATCHDIR("hive.blobstore.use.blobstore.as.scratchdir", false,
-            "Enable the use of scratch directories directly on blob storage systems (it may cause performance penalties).");
+            "Enable the use of scratch directories directly on blob storage systems (it may cause performance penalties)."),
+
+    HIVE_QUERY_TIMEOUT_SECONDS("hive.query.timeout.seconds", "0s",
+        "Timeout for Running Query in seconds. A nonpositive value means infinite. " +
+        "If the query timeout is also set by thrift API call, the smaller one will be taken.");
 
     public final String varname;
     private final String defaultExpr;
diff --git a/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java b/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java
index f01d289..abe0b9e 100644
--- a/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java
+++ b/service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java
@@ -110,6 +110,12 @@ public SQLOperation(HiveSession parentSession, String statement, Map<String, Str
     // TODO: call setRemoteUser in ExecuteStatementOperation or higher.
     super(parentSession, statement, confOverlay, runInBackground);
     this.queryTimeout = queryTimeout;
+    long timeout = HiveConf.getTimeVar(parentSession.getHiveConf(),
+        HiveConf.ConfVars.HIVE_QUERY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+    if (timeout > 0 && (queryTimeout <= 0 || timeout < queryTimeout)) {
+      this.queryTimeout = timeout;
+    }
+
     setupSessionIO(parentSession.getSessionState());
     try {
       sqlOpDisplay = new SQLOperationDisplay(this);
-- 
1.7.9.5

